File: go/search/query_parser.go
// SPDX-License-Identifier: Apache-2.0

package search

import (
	"fmt"
	"strings"
	"unicode"

	"golang.org/x/text/unicode/norm"
)

// TokenKind is the shape of a single unit produced by [ParseQuery].
type TokenKind int

const (
	// TokTerm is a bare word that should go through the FTS5
	// stemmer.
	TokTerm TokenKind = iota
	// TokPhrase is a multi-word string that should match exactly
	// (wrapped in double quotes on the FTS5 side).
	TokPhrase
	// TokPrefix is a word followed by `*` for prefix matching.
	TokPrefix
)

// String returns the spec name for the token kind. Mirrors the
// `kind` field on the AST nodes in spec/fixtures/query-parser/cases.json
// so the golden test can compare directly.
func (k TokenKind) String() string {
	switch k {
	case TokPhrase:
		return "phrase"
	case TokPrefix:
		return "prefix"
	default:
		return "term"
	}
}

// Token is one unit produced by [ParseQuery]. Operator is only
// populated when the user wrote an explicit uppercase boolean
// (`AND`/`OR`/`NOT`) immediately before the token; otherwise the
// default operator configured in [BuildFTS5Expr] applies.
type Token struct {
	Kind     TokenKind
	Text     string
	Operator string
}

// AST is the full parse result. Raw is the NFC + whitespace-collapsed
// form of the input. HasOperators records whether the input carried
// any quote or explicit uppercase AND/OR/NOT; per spec/QUERY-DSL.md
// that flag disables stopword filtering.
type AST struct {
	Raw          string
	Tokens       []Token
	HasOperators bool
}

// ftsTermReplacer strips FTS5 special characters and natural-language
// punctuation from bare terms so [BuildFTS5Expr] never emits a
// malformed expression. Phrases are exempt: their content is wrapped
// in double quotes, which FTS5 treats as a literal phrase. The set
// mirrors spec/QUERY-DSL.md ("Token stripping" section).
var ftsTermReplacer = strings.NewReplacer(
	"*", "",
	"(", "",
	")", "",
	":", "",
	"^", "",
	"+", "",
	`"`, "",
	"-", "",
	"?", "",
	"!", "",
	".", "",
	",", "",
	";", "",
	"/", "",
	"\\", "",
	"[", "",
	"]", "",
	"{", "",
	"}", "",
	"<", "",
	">", "",
	"|", "",
	"&", "",
	"'", "",
	"$", "",
	"#", "",
	"@", "",
	"%", "",
	"=", "",
	"~", "",
	"`", "",
)

// ftsPhraseReplacer scrubs the small number of characters that would
// prematurely close a phrase or break the FTS5 parser inside quoted
// content.
var ftsPhraseReplacer = strings.NewReplacer(
	`"`, "",
)

// normaliseInput applies the spec pre-parse normalisation: Unicode
// NFC, strip zero-width and BOM characters, map NBSP to regular
// space, collapse whitespace, trim. Returns the canonical `raw` form
// that the golden fixture compares against.
func normaliseInput(raw string) string {
	if raw == "" {
		return ""
	}
	n := norm.NFC.String(raw)
	var b strings.Builder
	for _, r := range n {
		switch r {
		case '\u200B', '\u200C', '\u200D', '\uFEFF':
			// Strip zero-width joiners / non-joiners / BOM.
			continue
		case '\u00A0':
			// NBSP mapped to regular space.
			b.WriteByte(' ')
		default:
			b.WriteRune(r)
		}
	}
	// Collapse runs of whitespace and trim.
	return strings.Join(strings.Fields(b.String()), " ")
}

// hasOperatorSignals reports whether the raw input contains at least
// one double quote or at least one whitespace-bounded uppercase
// boolean operator. When true the parser disables stopword filtering,
// preserving power-user intent verbatim.
func hasOperatorSignals(raw string) bool {
	if strings.ContainsRune(raw, '"') {
		return true
	}
	for _, field := range strings.Fields(raw) {
		if field == "AND" || field == "OR" || field == "NOT" {
			return true
		}
	}
	return false
}

// Parse walks raw and produces the full [AST]: the normalised `raw`
// string, the token stream, and the `hasOperators` flag documented in
// spec/QUERY-DSL.md.
//
// TODO(spec): jeff's upstream parser (apps/jeff/internal/search)
// always applied stopword filtering to bare terms regardless of
// whether the query contained quotes or explicit uppercase AND /
// OR / NOT. This port follows the spec's "disable stopword
// filtering when any operator or phrase is present" rule so the
// golden fixture passes; callers that relied on jeff's stricter
// filtering will see the stopword list survive in operator-bearing
// queries. The spec is the source of truth for cross-SDK
// determinism.
func Parse(raw string) AST {
	normalised := normaliseInput(raw)
	ast := AST{Raw: normalised}
	if normalised == "" {
		return ast
	}

	ast.HasOperators = hasOperatorSignals(normalised)
	filterStopwords := !ast.HasOperators

	runes := []rune(normalised)
	i := 0
	pendingOp := ""

	for i < len(runes) {
		r := runes[i]

		// Skip whitespace.
		if unicode.IsSpace(r) {
			i++
			continue
		}

		// Quoted phrase.
		if r == '"' {
			i++ // consume opening quote
			start := i
			for i < len(runes) && runes[i] != '"' {
				i++
			}
			phrase := string(runes[start:i])
			if i < len(runes) {
				i++ // consume closing quote
			}
			phrase = strings.TrimSpace(ftsPhraseReplacer.Replace(phrase))
			phrase = strings.ToLower(phrase)
			if phrase != "" {
				ast.Tokens = append(ast.Tokens, Token{
					Kind:     TokPhrase,
					Text:     phrase,
					Operator: pendingOp,
				})
				pendingOp = ""
			}
			continue
		}

		// Bare word: run until whitespace or quote.
		start := i
		for i < len(runes) && !unicode.IsSpace(runes[i]) && runes[i] != '"' {
			i++
		}
		word := string(runes[start:i])
		if word == "" {
			continue
		}

		// Uppercase boolean operator outside a phrase.
		if word == "AND" || word == "OR" || word == "NOT" {
			pendingOp = word
			continue
		}

		// Prefix token: trailing `*` survives the scrub; everything
		// else is cleaned.
		isPrefix := strings.HasSuffix(word, "*")
		if isPrefix {
			word = strings.TrimSuffix(word, "*")
		}

		// Keep the pre-scrub form around so the alias map can be
		// consulted with the user's original surface token. The FTS5
		// scrub strips hyphens, which would otherwise turn `a-ware`
		// into `aware` before alias lookup and silently miss the
		// entry.
		preScrubLower := strings.ToLower(strings.TrimSpace(word))

		cleaned := ftsTermReplacer.Replace(word)
		cleaned = strings.TrimSpace(cleaned)
		if cleaned == "" {
			pendingOp = ""
			continue
		}

		lower := strings.ToLower(cleaned)
		if filterStopwords {
			// Lowercase `and`, `or`, `not` are natural-language
			// filler in bare-word queries per spec/QUERY-DSL.md.
			// Drop them before consulting the stopword set.
			if lower == "and" || lower == "or" || lower == "not" {
				pendingOp = ""
				continue
			}
			if isStopWord(lower) {
				pendingOp = ""
				continue
			}
		}

		kind := TokTerm
		if isPrefix {
			kind = TokPrefix
		}

		// Alias expansion only applies to bare terms: phrases are
		// literal, prefix tokens already cover a family of stems, and
		// aliasing either of them would silently break user intent.
		if kind == TokTerm {
			expanded := expandAlias(preScrubLower)
			if len(expanded) == 1 && preScrubLower != lower {
				expanded = expandAlias(lower)
			}
			if len(expanded) > 1 {
				emitted := 0
				seen := map[string]bool{}
				for _, alt := range expanded {
					altToken, ok := aliasToken(alt)
					if !ok {
						continue
					}
					dedupeKey := fmt.Sprintf("%d|%s", altToken.Kind, altToken.Text)
					if seen[dedupeKey] {
						continue
					}
					seen[dedupeKey] = true

					if emitted == 0 {
						altToken.Operator = pendingOp
					}
					ast.Tokens = append(ast.Tokens, altToken)
					emitted++
				}
				if emitted > 0 {
					pendingOp = ""
					continue
				}
			}
		}

		ast.Tokens = append(ast.Tokens, Token{
			Kind:     kind,
			Text:     lower,
			Operator: pendingOp,
		})
		pendingOp = ""
	}

	return ast
}

// ParseQuery is the jeff-style token-stream entry point kept for
// compatibility with the ported tests. Prefer [Parse] for the full
// AST shape.
func ParseQuery(raw string) []Token {
	return Parse(raw).Tokens
}

// expandAlias consults the package-wide alias map and returns the
// alternatives for token, lowercased and deduplicated. When no alias
// map is installed or the token has no match, the original token is
// returned as a single-element slice so callers can branch on length.
func expandAlias(token string) []string {
	m := getAliasMap()
	if m == nil {
		return []string{token}
	}
	return m.Expand(token)
}

// aliasToken normalises a single alias alternative into a Token
// ready for [BuildFTS5Expr]. Multi-word or hyphenated alternatives
// become phrase tokens so the FTS5 tokenizer (porter+unicode61,
// which splits on non-alphanumeric) still matches them against the
// stored documents. Single-word alternatives stay as bare terms so
// they benefit from the stemmer. Returns (zero, false) when the
// alternative collapses to nothing after trimming.
func aliasToken(alt string) (Token, bool) {
	alt = strings.ToLower(strings.TrimSpace(alt))
	if alt == "" {
		return Token{}, false
	}

	normalised := strings.Map(func(r rune) rune {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			return r
		}
		return ' '
	}, alt)
	fields := strings.Fields(normalised)
	if len(fields) == 0 {
		return Token{}, false
	}
	if len(fields) == 1 {
		return Token{Kind: TokTerm, Text: fields[0]}, true
	}
	return Token{Kind: TokPhrase, Text: strings.Join(fields, " ")}, true
}

// BuildFTS5Expr converts parsed tokens into a valid FTS5 MATCH
// expression. Default operator between bare terms is `OR`. Explicit
// operators from the user override the default. Phrase tokens are
// wrapped in double quotes (the one place quoting is legitimate in
// the new pipeline). Prefix tokens keep their trailing `*`.
//
// FTS5 does not support `OR NOT`, so any `NOT` token preceded by the
// default `OR` is rewritten to `AND NOT`. Likewise an explicit `NOT`
// operator is always emitted as `AND NOT` when it is not the first
// token in the expression. A leading `NOT` is illegal in FTS5; the
// compiler drops the operator and keeps the operand.
func BuildFTS5Expr(tokens []Token) string {
	if len(tokens) == 0 {
		return ""
	}

	var b strings.Builder
	for _, tok := range tokens {
		piece := renderToken(tok)
		if piece == "" {
			continue
		}

		if b.Len() == 0 {
			// First token: an explicit leading NOT is not valid in
			// FTS5, so drop the operator prefix.
			b.WriteString(piece)
			continue
		}

		op := tok.Operator
		if op == "" {
			op = "OR"
		}
		if op == "NOT" {
			// Rewrite bare `NOT` into `AND NOT` because FTS5 only
			// accepts the latter after the first operand.
			op = "AND NOT"
		}

		b.WriteString(" ")
		b.WriteString(op)
		b.WriteString(" ")
		b.WriteString(piece)
	}

	return strings.TrimSpace(b.String())
}

// renderToken produces the FTS5 surface form for a single [Token].
// Phrase tokens are wrapped in double quotes; prefix tokens keep
// their trailing `*`. Returns an empty string for pathological
// tokens (e.g. whitespace-only phrases) so [BuildFTS5Expr] can skip
// them cleanly.
func renderToken(tok Token) string {
	text := strings.TrimSpace(tok.Text)
	if text == "" {
		return ""
	}
	switch tok.Kind {
	case TokPhrase:
		return `"` + text + `"`
	case TokPrefix:
		return text + "*"
	default:
		return text
	}
}

File: sdks/py/src/jeffs_brain_memory/search/query_parser.py
# SPDX-License-Identifier: Apache-2.0
"""Query DSL parser and FTS5 compiler.

Ported from ``go/search/query_parser.go`` and validated against
the cross-SDK goldens at ``spec/fixtures/query-parser/cases.json``.
Behaviour is bit-for-bit compatible with the TypeScript and Go SDKs:

- Parse: NFC-normalise, strip zero-width / BOM, collapse whitespace,
  tokenise into term / phrase / prefix nodes, drop stopwords when no
  operator signal is present.
- Compile: render tokens to an FTS5 ``MATCH`` expression. Default
  operator is ``OR``; explicit operators override; ``NOT`` rewrites to
  ``AND NOT`` after the first operand; a leading ``NOT`` is dropped.
"""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass, field
from typing import Literal

from .aliases import AliasTable
from .stopwords import is_stopword

__all__ = [
    "TokenKind",
    "Token",
    "QueryAST",
    "parse",
    "compile",
    "compile_fts",
    "expand_aliases",
    "sanitise_query",
    "FTSExpr",
    "strongest_term",
]

TokenKind = Literal["term", "phrase", "prefix"]
"""The shape of a single parsed unit."""

FTSExpr = str
"""An FTS5 ``MATCH`` expression produced by :func:`compile`."""

# Characters stripped from bare terms by the FTS5 cleaner. Mirrors
# ``ftsTermReplacer`` in the Go SDK and the ``Token stripping`` block
# of ``spec/QUERY-DSL.md``. ``*`` is stripped here too; a trailing
# ``*`` is captured as a prefix marker before cleaning runs.
_FTS_TERM_STRIP = str.maketrans(
    "",
    "",
    "*():^+\"-?!.,;/\\[]{}<>|&'$#@%=~`",
)

# Characters stripped from phrase text. The only offender is the
# double quote itself; FTS5 treats the phrase body as literal.
_FTS_PHRASE_STRIP = str.maketrans("", "", '"')

# Zero-width and BOM characters removed during normalisation.
_ZW_CODEPOINTS = frozenset(("\u200b", "\u200c", "\u200d", "\ufeff"))


@dataclass(frozen=True, slots=True)
class Token:
    """One unit produced by :func:`parse`.

    ``operator`` is non-empty only when the user wrote an explicit
    uppercase boolean immediately before the token.
    """

    kind: TokenKind
    text: str
    operator: str = ""


@dataclass(slots=True)
class QueryAST:
    """Full parse result.

    ``raw`` is the NFC + whitespace-collapsed form of the input.
    ``has_operators`` records whether the input carried any quote or
    whitespace-bounded uppercase ``AND`` / ``OR`` / ``NOT``; per spec
    that flag disables stopword filtering.
    """

    raw: str
    tokens: list[Token] = field(default_factory=list)
    has_operators: bool = False


def _normalise_input(raw: str) -> str:
    """Apply the spec pre-parse normalisation.

    Unicode NFC, strip zero-width and BOM characters, map NBSP to a
    regular space, collapse runs of whitespace, trim.
    """
    if not raw:
        return ""
    nfc = unicodedata.normalize("NFC", raw)
    buf: list[str] = []
    for ch in nfc:
        if ch in _ZW_CODEPOINTS:
            continue
        if ch == "\u00a0":
            buf.append(" ")
            continue
        buf.append(ch)
    # ``str.split`` with no argument collapses any Unicode whitespace
    # run into a single split, matching Go's ``strings.Fields``.
    return " ".join("".join(buf).split())


def _has_operator_signals(raw: str) -> bool:
    """Detect whether the query carries a quote or explicit operator.

    Matches the Go ``hasOperatorSignals`` helper: any ``"`` or any
    whitespace-bounded uppercase ``AND`` / ``OR`` / ``NOT`` disables
    stopword filtering for the whole query.
    """
    if '"' in raw:
        return True
    return any(field in {"AND", "OR", "NOT"} for field in raw.split())


def _alias_token(alt: str) -> Token | None:
    """Normalise a single alias alternative into a :class:`Token`.

    Single-word alternatives become bare terms; multi-word or
    punctuation-bearing alternatives become phrase tokens (joined with
    a single space) so the FTS5 ``porter unicode61`` tokeniser still
    matches them against stored documents. ``None`` is returned when
    the alternative collapses to nothing after trimming.
    """
    alt = alt.strip().lower()
    if not alt:
        return None
    buf: list[str] = []
    for ch in alt:
        if ch.isalpha() or ch.isdigit():
            buf.append(ch)
        else:
            buf.append(" ")
    parts = "".join(buf).split()
    if not parts:
        return None
    if len(parts) == 1:
        return Token(kind="term", text=parts[0])
    return Token(kind="phrase", text=" ".join(parts))


def parse(query: str, *, aliases: AliasTable | None = None) -> QueryAST:
    """Parse ``query`` into a :class:`QueryAST`.

    ``aliases`` is consulted inline for bare terms only: phrases and
    prefixes bypass alias expansion per ``spec/QUERY-DSL.md``.
    """
    normalised = _normalise_input(query)
    ast = QueryAST(raw=normalised)
    if not normalised:
        return ast

    ast.has_operators = _has_operator_signals(normalised)
    filter_stopwords = not ast.has_operators

    chars = list(normalised)
    i = 0
    n = len(chars)
    pending_op = ""

    while i < n:
        ch = chars[i]

        if ch.isspace():
            i += 1
            continue

        if ch == '"':
            i += 1  # consume opening quote
            start = i
            while i < n and chars[i] != '"':
                i += 1
            phrase = "".join(chars[start:i])
            if i < n:
                i += 1  # consume closing quote
            phrase = phrase.translate(_FTS_PHRASE_STRIP).strip().lower()
            if phrase:
                ast.tokens.append(Token(kind="phrase", text=phrase, operator=pending_op))
                pending_op = ""
            continue

        # Bare word: run until whitespace or quote.
        start = i
        while i < n and not chars[i].isspace() and chars[i] != '"':
            i += 1
        word = "".join(chars[start:i])
        if not word:
            continue

        if word in ("AND", "OR", "NOT"):
            pending_op = word
            continue

        is_prefix = word.endswith("*")
        if is_prefix:
            word = word[:-1]

        pre_scrub_lower = word.strip().lower()

        cleaned = word.translate(_FTS_TERM_STRIP).strip()
        if not cleaned:
            pending_op = ""
            continue

        lower = cleaned.lower()

        if filter_stopwords:
            # Lowercase AND / OR / NOT are natural-language filler in
            # bare queries.
            if lower in ("and", "or", "not"):
                pending_op = ""
                continue
            if is_stopword(lower):
                pending_op = ""
                continue

        kind: TokenKind = "prefix" if is_prefix else "term"

        if kind == "term" and aliases is not None:
            expanded = aliases.expand(pre_scrub_lower)
            if len(expanded) == 1 and pre_scrub_lower != lower:
                expanded = aliases.expand(lower)
            if len(expanded) > 1:
                emitted = 0
                seen: set[tuple[str, str]] = set()
                for alt in expanded:
                    alt_tok = _alias_token(alt)
                    if alt_tok is None:
                        continue
                    dedupe_key = (alt_tok.kind, alt_tok.text)
                    if dedupe_key in seen:
                        continue
                    seen.add(dedupe_key)
                    if emitted == 0:
                        alt_tok = Token(
                            kind=alt_tok.kind,
                            text=alt_tok.text,
                            operator=pending_op,
                        )
                    ast.tokens.append(alt_tok)
                    emitted += 1
                if emitted > 0:
                    pending_op = ""
                    continue

        ast.tokens.append(Token(kind=kind, text=lower, operator=pending_op))
        pending_op = ""

    return ast


def _render_token(tok: Token) -> str:
    """Produce the FTS5 surface form for ``tok``.

    Empty-text tokens return ``""`` so :func:`compile` can skip them
    cleanly.
    """
    text = tok.text.strip()
    if not text:
        return ""
    if tok.kind == "phrase":
        return f'"{text}"'
    if tok.kind == "prefix":
        return f"{text}*"
    return text


def compile(ast: QueryAST) -> FTSExpr:  # noqa: A001 - parity with TS/Go API
    """Compile ``ast.tokens`` into an FTS5 ``MATCH`` expression.

    Default connective between tokens is ``OR``. Explicit operators
    on individual tokens override the default. ``NOT`` rewrites to
    ``AND NOT`` after the first operand. A leading ``NOT`` is illegal
    in FTS5 so the operator is dropped for the first token.
    """
    parts: list[str] = []
    for tok in ast.tokens:
        piece = _render_token(tok)
        if not piece:
            continue

        if not parts:
            parts.append(piece)
            continue

        op = tok.operator or "OR"
        if op == "NOT":
            op = "AND NOT"

        parts.append(op)
        parts.append(piece)

    return " ".join(parts).strip()


def compile_fts(tokens: list[Token]) -> FTSExpr:
    """Legacy entry-point accepting a raw token slice.

    Kept for symmetry with the Go ``BuildFTS5Expr`` API used by the
    retrieval layer. Prefer :func:`compile` on a :class:`QueryAST`.
    """
    return compile(QueryAST(raw="", tokens=list(tokens)))


def expand_aliases(ast: QueryAST, aliases: AliasTable) -> QueryAST:
    """Re-expand an already-parsed AST against ``aliases``.

    Most callers pass ``aliases`` directly to :func:`parse`. This
    helper exists for retrieval pipelines that receive an AST from an
    upstream source and want to attach aliases post-hoc without
    re-parsing the raw text.
    """
    out = QueryAST(raw=ast.raw, has_operators=ast.has_operators)
    for tok in ast.tokens:
        if tok.kind != "term":
            out.tokens.append(tok)
            continue
        expanded = aliases.expand(tok.text)
        if len(expanded) <= 1:
            out.tokens.append(tok)
            continue
        emitted = 0
        seen: set[tuple[str, str]] = set()
        for alt in expanded:
            alt_tok = _alias_token(alt)
            if alt_tok is None:
                continue
            dedupe_key = (alt_tok.kind, alt_tok.text)
            if dedupe_key in seen:
                continue
            seen.add(dedupe_key)
            if emitted == 0:
                alt_tok = Token(kind=alt_tok.kind, text=alt_tok.text, operator=tok.operator)
            out.tokens.append(alt_tok)
            emitted += 1
        if emitted == 0:
            out.tokens.append(tok)
    return out


def sanitise_query(query: str) -> FTSExpr:
    """Convert a natural-language query into a valid FTS5 expression.

    Empty input or an AST with zero surviving tokens produces a light
    fallback: strip control punctuation and join the remaining words
    with a space so the caller still sees something search-worthy.
    """
    query = query.strip()
    if not query:
        return ""
    ast = parse(query)
    if not ast.tokens:
        return _fallback_sanitise(query)
    return compile(ast)


_FALLBACK_STRIP = str.maketrans("", "", "*():^+\"?!,;$#@%=")


def _fallback_sanitise(query: str) -> FTSExpr:
    stripped = query.translate(_FALLBACK_STRIP)
    words = stripped.split()
    return " ".join(words)


def strongest_term(query: str) -> str:
    """Return the longest non-stopword token of length >= 3.

    Ties break on earliest position; the first occurrence survives.
    Used by the retry ladder rung 1 in ``spec/ALGORITHMS.md``.
    """
    sanitised = query.lower()
    buf: list[str] = []
    for ch in sanitised:
        buf.append(ch if (ch.isalnum() or ch.isspace()) else " ")
    tokens = "".join(buf).split()
    best = ""
    for tok in tokens:
        if len(tok) < 3:
            continue
        if is_stopword(tok):
            continue
        if len(tok) > len(best):
            best = tok
    return best
